### PR TITLE
Feature batch based search

### PR DESCRIPTION
**Pull request by Davknapp**
Changed the search algorithm to a batch-based approach. Instead of passing queries one by one all queries and an array of the indices of all active_queries are passed to the search callback. This way we can make use of our other batch-based algorithms and also save a lot of function calls. 
Comparing the search tutorial I observed a massive performance increase (test with 200 000 particles on a single process, average of 3 runs):
current: 13,904914s
batch_based: 0,321643s
An additional test with 2 million particles run in 3,047s. 

A further improvement could be done by splitting the search and query-callback functions into two. I don't know why they are used for the same and I think it is a good idea to split them into two different callbacks. I am up for discussion, but I think it is a good idea to introduce this change in another PR if we decide to do so. 

**_All these boxes must be checked by the reviewers before merging the pull request:_**

As a reviewer please read through all the code lines and make sure that the code is fully understood, bug free, well-documented and well-structured.

#### General
- [ ] The reviewer executed the new code features at least once and checked the results manually

- [ ] The code follows the [t8code coding guidelines](https://github.com/holke/t8code/wiki/Coding-Guideline)
- [ ] New source/header files are properly added to the Makefiles
- [ ] The code is well documented
- [ ] All function declarations, structs/classes and their members have a proper doxygen documentation
- [ ] All new algorithms and data structures are sufficiently optimal in terms of memory and runtime (If this should be merged, but there is still potential for optimization, create a new issue)

#### Tests
- [ ] The code is covered in an existing or new test case using Google Test

#### Github action

- [ ] The code compiles without warning in debugging and release mode, with and without MPI (this should be executed automatically in a github action)
- [ ] All tests pass (in various configurations, this should be executed automatically in a github action)

  If the Pull request introduces code that is not covered by the github action (for example coupling with a new library):
  - [ ] Should this use case be added to the github action?
  - [ ] If not, does the specific use case compile and all tests pass (check manually)

#### Scripts and Wiki

- [ ] If a new directory with source-files is added, it must be covered by the `script/find_all_source_files.scp` to check the indentation of these files.
- [ ] If this PR introduces a new feature, it must be covered in an example/tutorial and a Wiki article.

#### Licence

- [ ] The author added a BSD statement to `doc/` (or already has one)

### CODE REVIEW

**Comment by Davknapp** [comment]
sorry, I mixed this PR with older commits. Will clean it up soon, but it is to late for it now ;)

**Comment by cburstedde** [comment]
Impressive speedup. We'll take a look and see if this makes any difference in p4est.

File: src/t8_forest/t8_forest_iterate.h [hunk 1]
```diff
@@ -37,24 +37,25 @@ typedef int (*t8_forest_iterate_face_fn) (t8_forest_t forest, t8_locidx_t ltreei
...
+ * \param[in] leaf_elements       the leaf elements in \a forest that are descendants of \a element (or the element 
+ *                                itself if \a is_leaf is true)
+ * \param[in] tree_leaf_index     the local index of the first leaf in \a leaf_elements
+ * \param[in] queries             if not NULL, a query that is passed through from the search function
+ * \param[in] query_indices       if \a query is not NULL the indices of \a query in the \a queries array from \ref 
+ *                                t8_forest_search
+ * \param[in, out] query_matches  if \a query is not NULL: true at the i-th index if and only if the element 'matches' 
+ *                                the query of the i-th query index. 
+ *                                if \a query is NULL: true if and only if the search should continue with the
+ *                                children of \a element and the queries should be performed for this element.
```
Comment (jmark) [inline src/t8_forest/t8_forest_iterate.h:53]: Please add documentation for the `num_active_querries` argument.

File: test/t8_forest/t8_gtest_search.cxx [hunk 2]
```diff
@@ -89,28 +90,34 @@ t8_test_search_all_fn (t8_forest_t forest, t8_locidx_t ltreeid, const t8_element
...
+  for (size_t iquery = 0; iquery < num_active_queries; iquery++) {
+    void *query = sc_array_index_int ((sc_array_t *) queries, iquery);
+    /* The query callback is always called with a query */
+    EXPECT_TRUE (query != NULL) << "query " << iquery << " is NULL.";
+    /* The query is an int with value 42 (see below) */
+    EXPECT_EQ (*(int *) query, 42) << "Wrong query argument passed to query callback.";
+    /* The query index gives the position of the query in the queries array
+    * of the calling search forest_search. Since there is only one query in the
+    * array in this test, the index must always be 0. */
+    //EXPECT_EQ ((int) query_index, 0) << "Wrong query index passed to query callback.";
```
Comment (jmark) [inline test/t8_forest/t8_gtest_search.cxx:107]: Forgot to delete this line?

File: tutorials/general/t8_tutorial_search.cxx [hunk 3]
```diff
@@ -190,27 +202,28 @@ t8_tutorial_search_query_callback (t8_forest_t forest, t8_locidx_t ltreeid, cons
...
   T8_ASSERT (query != NULL);
 
-  /* Test whether this particle is inside this element. */
-  particle_is_inside_element
-    = t8_forest_element_point_inside (forest, ltreeid, element, particle->coordinates, tolerance);
-  if (particle_is_inside_element) {
-    if (is_leaf) {
-      /* The particle is inside and this element is a leaf element.
+  /* Test whether the particles are inside this element. */
+  t8_forest_element_point_batch_inside (forest, ltreeid, element, coords, num_active_queries, query_matches, tolerance);
```
Comment (jmark) [inline tutorials/general/t8_tutorial_search.cxx:206]: Note, this function got renamed in the other PR, right? Might cause a merge conflict one way or another.

File: tutorials/general/t8_tutorial_search.cxx [hunk 4]
```diff
@@ -190,27 +202,28 @@ t8_tutorial_search_query_callback (t8_forest_t forest, t8_locidx_t ltreeid, cons
...
+  T8_FREE (coords);
+  for (size_t matches_id = 0; matches_id < num_active_queries; matches_id++) {
+    if (query_matches[matches_id]) {
+      if (is_leaf) {
+        /* The particle is inside and this element is a leaf element.
        * We mark the particle for being inside the partition and we increase
        * the particles_per_element counter of this element. */
-      /* In order to find the index of the element inside the array, we compute the
+        /* In order to find the index of the element inside the array, we compute the
```
Comment (jmark) [inline tutorials/general/t8_tutorial_search.cxx:214]: Looks like a misaligned comment.

File: tutorials/general/t8_tutorial_search.cxx [hunk 5]
```diff
@@ -190,27 +202,28 @@ t8_tutorial_search_query_callback (t8_forest_t forest, t8_locidx_t ltreeid, cons
...
+        size_t particle_id = *(size_t *) sc_array_index_int (query_indices, matches_id);
+        t8_tutorial_search_particle_t *particle
+          = (t8_tutorial_search_particle_t *) sc_array_index ((sc_array_t *) query, particle_id);
+        particle->is_inside_partition = 1;
+        *(double *) t8_sc_array_index_locidx (particles_per_element, element_index) += 1;
+      }
     }
-    /* The particles is inside the element. This query should remain active.
-     * If this element is not a leaf the search will continue with its children. */
-    return 1;
```
Comment (jmark) [inline tutorials/general/t8_tutorial_search.cxx:210]: Is this logic not needed anymore?

**Review by jmark — CHANGES_REQUESTED** [review]

File: tutorials/general/t8_tutorial_search.cxx (hunk 3 above)
Comment (Davknapp) [reply tutorials/general/t8_tutorial_search.cxx:206]: Yes, but as currently none of them is merged I will wait for the first one to merge and adapt the other one accordingly.

File: tutorials/general/t8_tutorial_search.cxx (hunk 5 above)
Comment (Davknapp) [reply tutorials/general/t8_tutorial_search.cxx:210]: No, this logic has been moved into the search-function itself.

File: src/t8_forest/t8_forest_iterate.h [hunk 6]
```diff
@@ -37,24 +37,27 @@ typedef int (*t8_forest_iterate_face_fn) (t8_forest_t forest, t8_locidx_t ltreei
...
+ * \param[in]                     The number of currently active queries. Does not have to be equal to query->elem_count,
```
Comment (jmark) [inline src/t8_forest/t8_forest_iterate.h:54]:
```suggestion
 * \param[in] num_active_queries  The number of currently active queries. Does not have to be equal to query->elem_count,
```

**Review by jmark — CHANGES_REQUESTED** [review]
Nice work! Just fix this little typo and we're set!

**Review by jmark — APPROVED** [review]

**Pull request merged**